cpp/examples/tpch/utils.hpp

```diff
@@ -0,0 +1,486 @@
+                                             std::vector<cudf::size_type> right_on,
+                                             cudf::null_equality compare_nulls)
+{
+  CUDF_FUNC_RANGE();
+  auto oob_policy                                    = cudf::out_of_bounds_policy::DONT_CHECK;
+  auto const left_selected                           = left_input.select(left_on);
+  auto const right_selected                          = right_input.select(right_on);
+  auto const [left_join_indices, right_join_indices] = cudf::inner_join(
+    left_selected, right_selected, compare_nulls, rmm::mr::get_current_device_resource());
+
+  auto const left_indices_span  = cudf::device_span<cudf::size_type const>{*left_join_indices};
+  auto const right_indices_span = cudf::device_span<cudf::size_type const>{*right_join_indices};
+
+  auto const left_indices_col  = cudf::column_view{left_indices_span};
+  auto const right_indices_col = cudf::column_view{right_indices_span};
+
+  auto const left_result  = cudf::gather(left_input, left_indices_col, oob_policy);
+  auto const right_result = cudf::gather(right_input, right_indices_col, oob_policy);
+
+  auto joined_cols = left_result->release();
+  auto right_cols  = right_result->release();
+  joined_cols.insert(joined_cols.end(),
+                     std::make_move_iterator(right_cols.begin()),
+                     std::make_move_iterator(right_cols.end()));
+  return std::make_unique<cudf::table>(std::move(joined_cols));
+}
+
+/**
+ * @brief Apply an inner join operation to two tables
+ *
+ * @param left_input The left input table
+ * @param right_input The right input table
+ * @param left_on The columns to join on in the left table
+ * @param right_on The columns to join on in the right table
+ * @param compare_nulls The null equality policy
+ */
+std::unique_ptr<table_with_cols> apply_inner_join(
+  std::unique_ptr<table_with_cols>& left_input,
+  std::unique_ptr<table_with_cols>& right_input,
+  std::vector<std::string> left_on,
+  std::vector<std::string> right_on,
+  cudf::null_equality compare_nulls = cudf::null_equality::EQUAL)
+{
+  CUDF_FUNC_RANGE();
+  std::vector<cudf::size_type> left_on_indices;
+  std::vector<cudf::size_type> right_on_indices;
+  for (auto& col_name : left_on) {
+    left_on_indices.push_back(left_input->col_id(col_name));
+  }
+  for (auto& col_name : right_on) {
+    right_on_indices.push_back(right_input->col_id(col_name));
+  }
+  auto table = join_and_gather(
+    left_input->table(), right_input->table(), left_on_indices, right_on_indices, compare_nulls);
+  return std::make_unique<table_with_cols>(std::move(table),
+                                           concat(left_input->columns(), right_input->columns()));
+}
+
+/**
+ * @brief Apply a filter predicated to a table
+ *
+ * @param table The input table
+ * @param predicate The filter predicate
+ */
+std::unique_ptr<table_with_cols> apply_filter(std::unique_ptr<table_with_cols>& table,
```

Comment:
```suggestion
std::unique_ptr<table_with_cols> apply_filter(std::unique_ptr<table_with_cols>const & table,
```
pass const ref, if this function does not modify this object.
If it modifies, pass non-const ref (minimize creating such functions).

Similar const ref comment for other functions as well.